Fix Email Authentication Issue Caused by Trailing Spaces

Mehul-Kumar-27 opened the pull request:
Fixes #518 

The issue can be easily resolved by applying the trim() function to the email and password strings before they are passed for authentication.

gnprice commented on .gitignore:
```diff
@@ -18,6 +18,7 @@ migrate_working_dir/
 /android/.idea/*
 !/.idea/inspectionProfiles/
 !/android/.idea/inspectionProfiles/
+.fvm
```
This seems potentially useful. It should go in its own commit, though, with its own commit message explaining the motivation for it. See the Zulip project style guide on clean commit history:
https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html

gnprice commented on pubspec.yaml:
```diff
@@ -24,8 +24,8 @@ environment:
   #   or pin exact versions: https://github.com/zulip/zulip-flutter/issues/15
-  sdk: '>=3.4.0-71.0.dev <4.0.0'
-  flutter: '>=3.19.0-9.0.pre.169'
+  sdk: '>=3.2.6 <4.0.0'
+  flutter: '>=3.16.9'
```
Don't downgrade our Flutter version :slightly_smiling_face: 

If this change works for you, it's OK to have locally; but I'd encourage you to use the Flutter main channel instead as described in our README file. In any case it's not a change we'd want to merge.

Mehul-Kumar-27 replied:
I appreciate your feedback and concerns. I'll stick to the Flutter main channel as per the README file.

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -1,4 +1,5 @@
 import 'dart:math';
+import 'dart:ui';
```
These added imports are failing CI because they're unnecessary. (I am a bit curious what caused them to appear in your editing, though.)

Mehul-Kumar-27 replied:
I was using Flutter version 3.16.9, and I imported certain files due to an error in the 'FontFeature.enable' part of the code, which was causing an unrecognized Target URI error. To resolve this, I had to import them.

However, after upgrading both Flutter and the SDK version, those errors are no longer present.

gnprice replied:
Ah, that makes sense. Yeah, so a good reason to match our Flutter version, then.

gnprice commented on lib/widgets/login.dart:
```diff
@@ -323,6 +323,7 @@ class _UsernamePasswordFormState extends State<_UsernamePasswordForm> {
   }
 
   void _submit() async {
+
```
nit: unrelated change; instead just leave this whitespace as it is

gnprice commented on lib/widgets/login.dart:
```diff
@@ -334,8 +335,8 @@ class _UsernamePasswordFormState extends State<_UsernamePasswordForm> {
       return;
     }
-    final String username = usernameFieldState.value!;
-    final String password = passwordFieldState.value!;
+    final String username = usernameFieldState.value!.trim();
+    final String password = passwordFieldState.value!.trim();
```
This is good for the username (aka the email, for Zulip realms where usernames are emails), but it seems less appropriate for the password. I think the usual practice is that if you put spaces in your password, then that's just part of the password like anything else.

gnprice reviewed (commented):
Thanks @Mehul-Kumar-27 for the contribution! It'll be good to fix this issue.

See comments below on some changes we'll need in order to make this a PR we can merge.

Normally I would also ask you to write a test for your change, but this function is actually one of a few spots where it's not currently possible to test without doing some refactoring. (There's a TODO comment to this effect inside the `fetchApiKey` function it calls.) I'll see if I can quick take care of that right now, so that it's possible for you to write a test.

gnprice commented:
> Normally I would also ask you to write a test for your change, but this function is actually one of a few spots where it's not currently possible to test without doing some refactoring. (There's a TODO comment to this effect inside the `fetchApiKey` function it calls.) I'll see if I can quick take care of that right now, so that it's possible for you to write a test.

OK, I've sent #522 with those changes. Please rebase your branch atop that PR, and then you should be able to write a test for this change: you can pretty much copy the `testWidgets` test that that PR adds, and modify what's entered into the username field with the `tester.enterText` call so that it exercises your new logic.

If you're not sure how to rebase your PR on top of another PR's branch, just ask in `#git help` on chat.zulip.org and we'll be glad to help out.

Mehul-Kumar-27 commented:
@gnprice , I have made the necessary changes, but somehow the CI is still failing. I am quite new to CI/CD. Could you please provide me with some context for the reason for the failure?

gnprice commented:
Thanks for the revision!

To fix errors that are appearing in CI, it's most convenient to run the same checks locally on your machine. You can do this with the command `tools/check`; see https://github.com/zulip/zulip-flutter/#tests in our README.

When the check script fails, it prints messages explaining what failed. But the version in CI prints a lot of other detail, so it can take a bit of practice to be able to find the relevant message in its output. If you run just `tools/check`, it's a lot more concise and you should be able to find the relevant complaint. (If you have any further trouble, though, please ask in `#mobile` on chat.zulip.org and we'll be glad to help you debug.)

gnprice commented:
> But the version in CI prints a lot of other detail, so it can take a bit of practice to be able to find the relevant message in its output.

Oh and the relevant piece of the GitHub web UI is a bit inconspicuous, too, so to make sure you've found it: there's a "Details" link here, on the right-hand side at the bottom of the PR thread:
![image](https://github.com/zulip/zulip-flutter/assets/28173/58c9b3fa-df47-4383-9321-377e246b3e20)

That links to the output of the latest CI run, which is currently this:
https://github.com/zulip/zulip-flutter/actions/runs/7919458559/job/21620278989?pr=519

And that's where you find the voluminous output I mentioned. (Which you may already have found.)

But as I said, probably the easiest way to diagnose the problem is to run `tools/check` locally. Once you've done so, it may be informative to compare your local `tools/check` output with that CI output — the relevant error message does appear in the CI version, so finding it there might help you find the error directly from the CI output next time.

gnprice commented on test/widgets/login_test.dart:
```diff
@@ -26,19 +37,112 @@ void main() {
+      await prepare(tester, serverSettings);
+      check(testBinding.globalStore.accounts).isEmpty();
+
+      final emailWithTrailingSpaces = '${eg.selfAccount.email}    ';
+      await tester.enterText(findUsernameInput, emailWithTrailingSpaces);
```
May as well add some spaces at the beginning as well as the end, to exercise both sides of what `String.trim` does.

gnprice commented on test/widgets/login_test.dart:
```diff
@@ -26,19 +37,112 @@ void main() {
+          .copyWith(id: testBinding.globalStore.accounts.single.id));
+    });
+
+    testWidgets('test case with trimming logic', (tester) async {
```
```suggestion
    testWidgets('trims whitespace on username', (tester) async {
```

gnprice reviewed (commented):
Thanks for the revision! The new test looks good, with a couple of small comments below.

The main thing needed now before we can merge this PR is to clean things up so that it doesn't make other changes that are unrelated to what it's about. Please take a look at the Zulip project's style guide for a clean commit history:
https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html

This PR should end up being a single commit on top of #522 (if the `.gitignore`  change is dropped, or two commits where one is the `.gitignore` change), just changing the one line to add `.trim()` and adding the test case, and not making any other changes. There also should be no merge commits.

To get a clear view of what changes your commits are making and of the structure of your branch, I recommend [a graphical Git viewer](https://zulip.readthedocs.io/en/latest/git/setup.html#get-a-graphical-client) like `gitk`. Alternatively, for reading in your terminal, `git log -p` is very helpful in combination with [my "secret"](https://github.com/zulip/zulip-mobile/blob/main/docs/howto/git.md#git-log-secret) for getting a good UI with it. If you have any questions, we'll be glad to answer in `#git help` on chat.zulip.org.

Mehul-Kumar-27 commented:
> Thanks for the revision! The new test looks good, with a couple of small comments below.
> 
> The main thing needed now before we can merge this PR is to clean things up so that it doesn't make other changes that are unrelated to what it's about. Please take a look at the Zulip project's style guide for a clean commit history: https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html
> 
> This PR should end up being a single commit on top of #522 (if the `.gitignore` change is dropped, or two commits where one is the `.gitignore` change), just changing the one line to add `.trim()` and adding the test case, and not making any other changes. There also should be no merge commits.
> 
> To get a clear view of what changes your commits are making and of the structure of your branch, I recommend [a graphical Git viewer](https://zulip.readthedocs.io/en/latest/git/setup.html#get-a-graphical-client) like `gitk`. Alternatively, for reading in your terminal, `git log -p` is very helpful in combination with [my "secret"](https://github.com/zulip/zulip-mobile/blob/main/docs/howto/git.md#git-log-secret) for getting a good UI with it. If you have any questions, we'll be glad to answer in `#git help` on chat.zulip.org.

Thank you for the suggestions, sir. I will definitely look into them and continue improving my contributions.

Mehul-Kumar-27 commented:
@gnprice , I've incorporated the suggested changes by adding white spaces at the start and modifying the test name. I've also addressed the CI failure caused by the unwanted import and have also removed the gitignore change. If there are any additional tasks, please let me know.

gnprice reviewed (commented):
Thanks. Two nits on the test, but the main things needed before we can merge this are the same as before: https://github.com/zulip/zulip-flutter/pull/519#pullrequestreview-1883589525

Here's the current structure of the branch, as shown by `gitk origin..`:
![image](https://github.com/zulip/zulip-flutter/assets/28173/8383aaaa-1490-4114-ae1e-bc0d1491a6f5)

So there's the commits that are part of #522, and then there are 10 further commits on top, one of which is a merge commit. There should instead be just one commit on top of #522.

And here's the total diff relative to #522:
```
$ git diff --stat 8be9833df
 .gitignore                          |  1 +
 lib/widgets/login.dart              |  3 +-
 lib/widgets/unread_count_badge.dart |  1 -
 pubspec.lock                        | 32 ++++++++--------
 test/flutter_checks.dart            |  1 +
 test/widgets/login_test.dart        | 80 ++++++++++++++++++++++++++--------------
 test/widgets/text_test.dart         |  1 +
 7 files changed, 74 insertions(+), 45 deletions(-)
```

You can say `git diff --stat -p` instead of `git diff --stat` and get the full details of the diff. Most of those changes are unrelated to what this PR is about, so should be left out. For details on some of them, see https://github.com/zulip/zulip-flutter/pull/519#discussion_r1491475819 and https://github.com/zulip/zulip-flutter/pull/519#discussion_r1490172130.

gnprice commented:
That other PR #522 has been merged, so if you rebase onto current main, that should help simplify the picture in Git. This PR should now be just one commit on top of main.

Mehul-Kumar-27 commented:
I am closing this and opening a new MR for this :)

gnprice commented:
Cool, thanks for the update. For the future, though, it'd be better to update the existing PR rather than close it and open a new one: https://github.com/zulip/zulip-flutter/pull/547#pullrequestreview-1918098359
